README.md
# ghrls(1)

[![GitHub Actions](https://github.com/dtan4/k8stail/workflows/Test/badge.svg)](https://github.com/dtan4/ghrls/actions?query=workflow%3ATest+branch%3Amaster)

List & Describe GitHub Releases

```bash
$ ghrls list kubernetes/kubernetes | head
TAG               TYPE           CREATEDAT                        NAME
v1.6.0-alpha.0    TAG
v1.5.3-beta.0     TAG
v1.5.2            TAG+RELEASE    2017-01-12 13:51:15 +0900 JST    v1.5.2
v1.5.2-beta.0     TAG
v1.5.1            TAG+RELEASE    2016-12-14 09:50:36 +0900 JST    v1.5.1
v1.5.1-beta.0     TAG
v1.5.0            TAG+RELEASE    2016-12-13 08:29:43 +0900 JST    v1.5.0
v1.5.0-beta.3     TAG+RELEASE    2016-12-09 06:52:35 +0900 JST    v1.5.0-beta.3
v1.5.0-beta.2     TAG+RELEASE    2016-11-25 07:29:04 +0900 JST    v1.5.0-beta.2
```

## Installation

### Using Homebrew (OS X only)

Formula is available at [dtan4/homebrew-tools](https://github.com/dtan4/homebrew-tools).

```bash
$ brew tap dtan4/tools
$ brew install ghrls
```

### Precompiled binary

Precompiled binaries for Windows, OS X, Linux are available at [Releases](https://github.com/dtan4/ghrls/releases).

### From source

```bash
$ go get -d github.com/dtan4/ghrls
$ cd $GOPATH/src/github.com/dtan4/ghrls
$ make install
```

## Usage

### (Recommended) Set `GITHUB_TOKEN` environment variable

`ghrls` can be used without any special token as default.
However, according to [GitHub API Rate Limiting](https://developer.github.com/v3/#rate-limiting), unauthenticated user is allowed to make up to 60 requests per hour.
In contrast, authenticated user (using `GITHUB_TOKEN`) can make up to 5,000 requests per hour.
Authentication is required to access to yourt private repositories, of course.

I strongly recommend to set `GITHUB_TOKEN` in your shell. `GITHUB_TOKEN` can be obtained from [here](https://github.com/settings/tokens).

```bash
export GITHUB_TOKEN=xxxxxxxxxxxxxxxxxxxx
```

### `ghrls get`

Describe release information

```bash
$ ghrls get kubernetes/kubernetes v1.5.2
Tag:         v1.5.2
Commit:      08e099554f3c31f6e6f07b448ab3ed78d0520507
Name:        v1.5.2
Author:      saad-ali
CreatedAt:   2017-01-12 13:51:15 +0900 JST
PublishedAt: 2017-01-12 16:25:50 +0900 JST
URL:         https://github.com/kubernetes/kubernetes/releases/tag/v1.5.2
Assets:      https://github.com/kubernetes/kubernetes/releases/download/v1.5.2/kubernetes.tar.gz

See [kubernetes-announce@](https://groups.google.com/forum/#!forum/kubernetes-announce) and [CHANGELOG](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG.md#v152) for details.

SHA256 for `kubernetes.tar.gz`: `67344958325a70348db5c4e35e59f9c3552232cdc34defb8a0a799ed91c671a3`

Additional binary downloads are linked in the [CHANGELOG](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG.md#downloads-for-v152).
```

### `ghrls list`

List releases


```bash
$ ghrls list kubernetes/kubernetes | head
TAG               TYPE           CREATEDAT                        NAME
v1.6.0-alpha.0    TAG
v1.5.3-beta.0     TAG
v1.5.2            TAG+RELEASE    2017-01-12 13:51:15 +0900 JST    v1.5.2
v1.5.2-beta.0     TAG
v1.5.1            TAG+RELEASE    2016-12-14 09:50:36 +0900 JST    v1.5.1
v1.5.1-beta.0     TAG
v1.5.0            TAG+RELEASE    2016-12-13 08:29:43 +0900 JST    v1.5.0
v1.5.0-beta.3     TAG+RELEASE    2016-12-09 06:52:35 +0900 JST    v1.5.0-beta.3
v1.5.0-beta.2     TAG+RELEASE    2016-11-25 07:29:04 +0900 JST    v1.5.0-beta.2
```

## Development

Retrieve this repository and build using `make`.

```bash
$ go get -d github.com/dtan4/ghrls
$ cd $GOPATH/src/github.com/dtan4/ghrls
$ make
```

## Author

Daisuke Fujita ([@dtan4](https://github.com/dtan4))

## License

[![MIT License](http://img.shields.io/badge/license-MIT-blue.svg?style=flat)](LICENSE)


CHANGELOG.md
# [v0.2.1](https://github.com/dtan4/ghrls/releases/tag/v0.2.1) (2021-02-23)

- Correct embedded version information

# [v0.2.0](https://github.com/dtan4/ghrls/releases/tag/v0.2.0) (2021-02-23)

- Improve error handling ([#1](https://github.com/dtan4/ghrls/pull/1), thanks @b4b4r07)
- Use Go 1.16
- Update dependencies to the latest ones
  - See [`go.mod`](https://github.com/dtan4/ghrls/blob/dc204050c9796ff93da3d24beb752e803d719c08/go.mod)

# [v0.1.0](https://github.com/dtan4/ghrls/releases/tag/v0.1.0) (2017-01-20)

Initial release.


cmd/get_test.go
package cmd

import (
	"bytes"
	"context"
	"fmt"
	"testing"
	"time"

	"github.com/dtan4/ghrls/github"
)

type fakeClientForGet struct {
	Tag *github.Tag
	Err error
}

func (c fakeClientForGet) DescribeRelease(ctx context.Context, owner, repo, tag string) (*github.Tag, error) {
	if c.Err != nil {
		return nil, c.Err
	}

	return c.Tag, nil
}

func (c fakeClientForGet) ListTagsAndReleases(ctx context.Context, owner, repo string) ([]*github.Tag, error) {
	return []*github.Tag{}, nil
}

func TestRunTag_success(t *testing.T) {
	gmt, err := time.LoadLocation("Europe/London")
	if err != nil {
		t.Fatal(err)
	}

	jst, err := time.LoadLocation("Asia/Tokyo")
	if err != nil {
		t.Fatal(err)
	}

	testcases := []struct {
		args     []string
		timezone *time.Location
		want     string
	}{
		{
			args:     []string{"owner/repo", "v1.5.2"},
			timezone: gmt,
			want: "" +
				"Tag:         v1\n" +
				"Commit:      856abeb2b507fc1db16dcaea938775ff938a5355\n" +
				"Name:        v1\n" +
				"Author:      dtan4\n" +
				"CreatedAt:   2018-12-13 00:30:24 +0000 GMT\n" +
				"PublishedAt: 2018-12-14 00:30:24 +0000 GMT\n" +
				"URL:         https://github.com/owner/repo/releases/tag/v1\n" +
				"Artifacts:   https://github.com/owner/repo/releases/download/v1/darwin.tar.gz\n" +
				"\n" +
				"The quick brown fox jumps over the lazy dog\n",
		},
		{
			args:     []string{"owner/repo", "v1.5.2"},
			timezone: jst,
			want: "" +
				"Tag:         v1\n" +
				"Commit:      856abeb2b507fc1db16dcaea938775ff938a5355\n" +
				"Name:        v1\n" +
				"Author:      dtan4\n" +
				"CreatedAt:   2018-12-13 09:30:24 +0900 JST\n" +
				"PublishedAt: 2018-12-14 09:30:24 +0900 JST\n" +
				"URL:         https://github.com/owner/repo/releases/tag/v1\n" +
				"Artifacts:   https://github.com/owner/repo/releases/download/v1/darwin.tar.gz\n" +
				"\n" +
				"The quick brown fox jumps over the lazy dog\n",
		},
	}

	client := fakeClientForGet{
		Tag: &github.Tag{
			Name: "v1",
			Release: &github.Release{
				ArtifactURLs: []string{
					"https://github.com/owner/repo/releases/download/v1/darwin.tar.gz",
				},
				Author:      "dtan4",
				Body:        "The quick brown fox jumps over the lazy dog",
				Commit:      "856abeb2b507fc1db16dcaea938775ff938a5355",
				CreatedAt:   time.Date(2018, 12, 13, 0, 30, 24, 0, time.UTC),
				Name:        "v1",
				PublishedAt: time.Date(2018, 12, 14, 0, 30, 24, 0, time.UTC),
				URL:         "https://github.com/owner/repo/releases/tag/v1",
			},
		},
	}

	for _, tc := range testcases {
		stdout, stderr := new(bytes.Buffer), new(bytes.Buffer)

		if err := RunGet(stdout, stderr, tc.args, client, tc.timezone); err != nil {
			t.Errorf("want: no error, got: %#v", err)
		}

		if stdout.String() != tc.want {
			t.Errorf("stderr want:\n%q\ngot:\n%q", tc.want, stdout.String())
		}
	}
}

func TestRunGet_invalidArgs(t *testing.T) {
	testcases := []struct {
		args []string
		want string
	}{
		{
			args: []string{},
			want: "Please specify repository <user/name> and tag.",
		},
		{
			args: []string{"owner/repo"},
			want: "Please specify repository <user/name> and tag.",
		},
		{
			args: []string{"owner/repo", "v1", "foobar"},
			want: "Please specify repository <user/name> and tag.",
		},
		{
			args: []string{"owner repo", "v1"},
			want: "Invalid repository name: owner repo",
		},
	}

	gmt, err := time.LoadLocation("Europe/London")
	if err != nil {
		t.Fatal(err)
	}

	for _, tc := range testcases {
		stdout, stderr := new(bytes.Buffer), new(bytes.Buffer)

		err := RunGet(stdout, stderr, tc.args, fakeClientForGet{}, gmt)

		if err == nil {
			t.Error("want: error, got: nil")
		}

		if err.Error() != tc.want {
			t.Errorf("error want: %q, got: %q", tc.want, err.Error())
		}
	}
}

func TestRunGet_notfound(t *testing.T) {
	testcases := []struct {
		args []string
		err  error
		want string
	}{
		{
			args: []string{"owner/repo", "v1"},
			err:  fmt.Errorf("404 Not Found"),
			want: "owner/repo@v1 : not found",
		},
		{
			args: []string{"owner/repo", "v1"},
			err:  fmt.Errorf("unexpected error"),
			want: "unexpected error",
		},
	}

	gmt, err := time.LoadLocation("Europe/London")
	if err != nil {
		t.Fatal(err)
	}

	for _, tc := range testcases {
		stdout, stderr := new(bytes.Buffer), new(bytes.Buffer)
		client := fakeClientForGet{
			Err: tc.err,
		}

		err := RunGet(stdout, stderr, tc.args, client, gmt)

		if err == nil {
			t.Error("want: error, got: nil")
		}

		if err.Error() != tc.want {
			t.Errorf("error want: %q, got: %q", tc.want, err.Error())
		}
	}
}


cmd/list_test.go
package cmd

import (
	"bytes"
	"context"
	"fmt"
	"testing"
	"time"

	"github.com/dtan4/ghrls/github"
)

type fakeClientForList struct {
	Tags []*github.Tag
	Err  error
}

func (c fakeClientForList) DescribeRelease(ctx context.Context, owner, repo, tag string) (*github.Tag, error) {
	return &github.Tag{}, nil
}

func (c fakeClientForList) ListTagsAndReleases(ctx context.Context, owner, repo string) ([]*github.Tag, error) {
	if c.Err != nil {
		return []*github.Tag{}, c.Err
	}

	return c.Tags, nil
}

func TestRunList_success(t *testing.T) {
	gmt, err := time.LoadLocation("Europe/London")
	if err != nil {
		t.Fatal(err)
	}

	jst, err := time.LoadLocation("Asia/Tokyo")
	if err != nil {
		t.Fatal(err)
	}

	testcases := []struct {
		args     []string
		timezone *time.Location
		want     string
	}{
		{
			args:     []string{"owner/repo"},
			timezone: gmt,
			want: "" +
				"TAG               TYPE           CREATEDAT                        NAME\n" +
				"v1.6.0-alpha.0    TAG                                             \n" +
				"v1.5.3-beta.0     TAG                                             \n" +
				"v1.5.2            TAG+RELEASE    2017-01-12 04:51:15 +0000 GMT    v1.5.2\n",
		},
		{
			args:     []string{"owner/repo"},
			timezone: jst,
			want: "" +
				"TAG               TYPE           CREATEDAT                        NAME\n" +
				"v1.6.0-alpha.0    TAG                                             \n" +
				"v1.5.3-beta.0     TAG                                             \n" +
				"v1.5.2            TAG+RELEASE    2017-01-12 13:51:15 +0900 JST    v1.5.2\n",
		},
	}

	client := fakeClientForList{
		Tags: []*github.Tag{
			&github.Tag{
				Name:    "v1.6.0-alpha.0",
				Release: nil,
			},
			&github.Tag{
				Name:    "v1.5.3-beta.0",
				Release: nil,
			},
			&github.Tag{
				Name: "v1.5.2",
				Release: &github.Release{
					CreatedAt: time.Date(2017, 1, 12, 4, 51, 15, 0, time.UTC),
					Name:      "v1.5.2",
				},
			},
		},
	}

	for _, tc := range testcases {
		stdout, stderr := new(bytes.Buffer), new(bytes.Buffer)

		if err := RunList(stdout, stderr, tc.args, client, tc.timezone); err != nil {
			t.Errorf("want: no error, got: %#v", err)
		}

		if stdout.String() != tc.want {
			t.Errorf("stderr want:\n%q\ngot:\n%q", tc.want, stdout.String())
		}
	}
}

func TestRunList_invalidArgs(t *testing.T) {
	testcases := []struct {
		args []string
		want string
	}{
		{
			args: []string{},
			want: "Please specify repository <user/name>.",
		},
		{
			args: []string{"owner", "repo"},
			want: "Please specify repository <user/name>.",
		},
		{
			args: []string{"owner repo"},
			want: "Invalid repository name: owner repo",
		},
	}

	gmt, err := time.LoadLocation("Europe/London")
	if err != nil {
		t.Fatal(err)
	}

	for _, tc := range testcases {
		stdout, stderr := new(bytes.Buffer), new(bytes.Buffer)

		err := RunList(stdout, stderr, tc.args, fakeClientForList{}, gmt)

		if err == nil {
			t.Error("want: error, got: nil")
		}

		if err.Error() != tc.want {
			t.Errorf("error want: %q, got: %q", tc.want, err.Error())
		}
	}
}

func TestRunList_notfound(t *testing.T) {
	testcases := []struct {
		args []string
		err  error
		want string
	}{
		{
			args: []string{"owner/repo"},
			err:  fmt.Errorf("404 Not Found"),
			want: "owner/repo: not found",
		},
		{
			args: []string{"owner/repo"},
			err:  fmt.Errorf("unexpected error"),
			want: "unexpected error",
		},
	}

	gmt, err := time.LoadLocation("Europe/London")
	if err != nil {
		t.Fatal(err)
	}

	for _, tc := range testcases {
		stdout, stderr := new(bytes.Buffer), new(bytes.Buffer)
		client := fakeClientForList{
			Err: tc.err,
		}

		err := RunList(stdout, stderr, tc.args, client, gmt)

		if err == nil {
			t.Error("want: error, got: nil")
		}

		if err.Error() != tc.want {
			t.Errorf("error want: %q, got: %q", tc.want, err.Error())
		}
	}
}


github/github_test.go
package github

import (
	"context"
	"reflect"
	"testing"
	"time"

	"github.com/google/go-github/v33/github"
)

type fakeRepositoriesService struct{}

func (s fakeRepositoriesService) GetReleaseByTag(ctx context.Context, owner, repo, tag string) (*github.RepositoryRelease, *github.Response, error) {
	tagName := "v1"
	body := "The quick brown fox jumps over the lazy dog"
	assetURL := "https://github.com/owner/repo/releases/download/v1/darwin.tar.gz"
	login := "dtan4"
	name := "v1"
	htmlURL := "https://github.com/owner/repo/releases/tag/v1"

	return &github.RepositoryRelease{
		Assets: []*github.ReleaseAsset{
			&github.ReleaseAsset{
				BrowserDownloadURL: &assetURL,
			},
		},
		Author: &github.User{
			Login: &login,
		},
		Body:        &body,
		CreatedAt:   &github.Timestamp{time.Date(2018, 12, 13, 0, 30, 24, 0, time.UTC)},
		HTMLURL:     &htmlURL,
		Name:        &name,
		PublishedAt: &github.Timestamp{time.Date(2018, 12, 14, 0, 30, 24, 0, time.UTC)},
		TagName:     &tagName,
	}, &github.Response{}, nil
}

func (s fakeRepositoriesService) GetCommit(ctx context.Context, owner, repo, sha string) (*github.RepositoryCommit, *github.Response, error) {
	commitSHA := "856abeb2b507fc1db16dcaea938775ff938a5355"

	return &github.RepositoryCommit{
		SHA: &commitSHA,
	}, &github.Response{}, nil
}

func (s fakeRepositoriesService) ListReleases(ctx context.Context, owner, repo string, opt *github.ListOptions) ([]*github.RepositoryRelease, *github.Response, error) {
	tag_v1_13_2_beta_0 := "v1.13.2-beta.0"
	tag_v1_13_1 := "v1.13.1"
	release_v1_13_1 := "v1.13.1"

	return []*github.RepositoryRelease{
		&github.RepositoryRelease{
			TagName:   &tag_v1_13_2_beta_0,
			Name:      nil,
			CreatedAt: &github.Timestamp{time.Date(2018, 12, 14, 0, 30, 24, 0, time.UTC)},
		},
		&github.RepositoryRelease{
			TagName:   &tag_v1_13_1,
			Name:      &release_v1_13_1,
			CreatedAt: &github.Timestamp{time.Date(2018, 12, 13, 0, 30, 24, 0, time.UTC)},
		},
	}, &github.Response{}, nil
}

func (s fakeRepositoriesService) ListTags(ctx context.Context, owner string, repo string, opt *github.ListOptions) ([]*github.RepositoryTag, *github.Response, error) {
	tag_v1_13_2_beta_1 := "v1.13.2-beta.1"
	tag_v1_13_2_beta_0 := "v1.13.2-beta.0"
	tag_v1_13_1 := "v1.13.1"

	return []*github.RepositoryTag{
		&github.RepositoryTag{
			Name: &tag_v1_13_2_beta_1,
		},
		&github.RepositoryTag{
			Name: &tag_v1_13_2_beta_0,
		},
		&github.RepositoryTag{
			Name: &tag_v1_13_1,
		},
	}, &github.Response{}, nil
}

func TestNewClient(t *testing.T) {
	testcases := []struct {
		accessToken string
	}{
		{accessToken: ""},
		{accessToken: "dummyaccesstoken"},
	}

	for _, tc := range testcases {
		c := NewClient(tc.accessToken)

		if c == nil {
			t.Error("want: object, got: nil")
		}
	}
}

func TestDescribeRelease(t *testing.T) {
	c := &Client{
		repositories: fakeRepositoriesService{},
	}

	owner := "owner"
	repo := "repo"
	tag := "v1"

	want := &Tag{
		Name: "v1",
		Release: &Release{
			ArtifactURLs: []string{
				"https://github.com/owner/repo/releases/download/v1/darwin.tar.gz",
			},
			Author:      "dtan4",
			Body:        "The quick brown fox jumps over the lazy dog",
			Commit:      "856abeb2b507fc1db16dcaea938775ff938a5355",
			CreatedAt:   time.Date(2018, 12, 13, 0, 30, 24, 0, time.UTC),
			Name:        "v1",
			PublishedAt: time.Date(2018, 12, 14, 0, 30, 24, 0, time.UTC),
			URL:         "https://github.com/owner/repo/releases/tag/v1",
		},
	}

	got, err := c.DescribeRelease(context.Background(), owner, repo, tag)
	if err != nil {
		t.Errorf("want no error, got: %#v", err)
	}

	if !reflect.DeepEqual(*got, *want) {
		t.Errorf("want: %#v, got: %#v", *want, *got)
	}
}

func TestListTagsAndReleases(t *testing.T) {
	c := &Client{
		repositories: fakeRepositoriesService{},
	}

	owner := "owner"
	repo := "repo"

	want := []*Tag{
		&Tag{
			Name:    "v1.13.2-beta.1",
			Release: nil,
		},
		&Tag{
			Name: "v1.13.2-beta.0",
			Release: &Release{
				Name:      "",
				CreatedAt: time.Date(2018, 12, 14, 0, 30, 24, 0, time.UTC),
			},
		},
		&Tag{
			Name: "v1.13.1",
			Release: &Release{
				Name:      "v1.13.1",
				CreatedAt: time.Date(2018, 12, 13, 0, 30, 24, 0, time.UTC),
			},
		},
	}

	got, err := c.ListTagsAndReleases(context.Background(), owner, repo)
	if err != nil {
		t.Errorf("want no error, got: %#v", err)
	}

	if len(got) != len(want) {
		t.Errorf("want: %d items, got: %d items", len(want), len(got))
	}

	for i, g := range got {
		if !reflect.DeepEqual(*g, *want[i]) {
			t.Errorf("want: %#v, got: %#v", *want[i], *g)
		}
	}
}


cmd/get.go
package cmd

import (
	"context"
	"fmt"
	"io"
	"os"
	"strings"
	"text/tabwriter"
	"time"

	"github.com/dtan4/ghrls/github"
	"github.com/spf13/cobra"
)

// getCmd represents the get command
var getCmd = &cobra.Command{
	Use:   "get REPOSITORY TAG",
	Short: "Describe release information",
	Long: `Describe release information

Example:

$ ghrls get kubernetes/kubernetes v1.5.2
Tag:         v1.5.2
Commit:      08e099554f3c31f6e6f07b448ab3ed78d0520507
Name:        v1.5.2
Author:      saad-ali
CreatedAt:   2017-01-12 13:51:15 +0900 JST
PublishedAt: 2017-01-12 16:25:50 +0900 JST
URL:         https://github.com/kubernetes/kubernetes/releases/tag/v1.5.2
Assets:      https://github.com/kubernetes/kubernetes/releases/download/v1.5.2/kubernetes.tar.gz

See [kubernetes-announce@](https://groups.google.com/forum/#!forum/kubernetes-announce) and [CHANGELOG](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG.md#v152) for details.
` +
		"SHA256 for `kubernetes.tar.gz`: `67344958325a70348db5c4e35e59f9c3552232cdc34defb8a0a799ed91c671a3`" +
		`
Additional binary downloads are linked in the [CHANGELOG](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG.md#downloads-for-v152).
`,
	RunE: func(cmd *cobra.Command, args []string) error {
		timezone := time.Local
		client := github.NewClient(rootOpts.GitHubToken)

		return RunGet(os.Stderr, os.Stdout, args, client, timezone)
	},
}

func RunGet(stdout, stderr io.Writer, args []string, client github.ClientInterface, timezone *time.Location) error {
	if len(args) != 2 {
		return fmt.Errorf("Please specify repository <user/name> and tag.")
	}

	ss := strings.Split(args[0], "/")
	if len(ss) != 2 {
		return fmt.Errorf("Invalid repository name: %s", args[0])
	}
	owner, repo := ss[0], ss[1]

	tag := args[1]

	ctx := context.Background()

	t, err := client.DescribeRelease(ctx, owner, repo, tag)
	if err != nil {
		if strings.Contains(err.Error(), "404 Not Found") {
			return fmt.Errorf("%s/%s@%s : not found", owner, repo, tag)
		}
		return err
	}

	w := tabwriter.NewWriter(stdout, 0, 0, 1, ' ', 0)
	fmt.Fprintln(w, "Tag:\t"+t.Name)
	fmt.Fprintln(w, "Commit:\t"+t.Release.Commit)
	fmt.Fprintln(w, "Name:\t"+t.Release.Name)
	fmt.Fprintln(w, "Author:\t"+t.Release.Author)
	fmt.Fprintln(w, "CreatedAt:\t"+t.Release.CreatedAt.In(timezone).String())
	fmt.Fprintln(w, "PublishedAt:\t"+t.Release.PublishedAt.In(timezone).String())
	fmt.Fprintln(w, "URL:\t"+t.Release.URL)

	if len(t.Release.ArtifactURLs) > 0 {
		fmt.Fprintln(w, "Artifacts:\t"+t.Release.ArtifactURLs[0])

		for _, url := range t.Release.ArtifactURLs[1:] {
			fmt.Fprintln(w, "\t"+url)
		}
	}

	w.Flush()

	if t.Release.Body != "" {
		fmt.Fprintln(stdout, "")
		fmt.Fprintln(stdout, t.Release.Body)
	}

	return nil
}

func init() {
	RootCmd.AddCommand(getCmd)
}


cmd/root.go
package cmd

import (
	"fmt"
	"os"

	"github.com/spf13/cobra"
)

// RootCmd represents the base command when called without any subcommands
var RootCmd = &cobra.Command{
	SilenceErrors: true,
	SilenceUsage:  true,
	Use:           "ghrls",
	Short:         "List & Describe GitHub Releases",
}

var rootOpts = struct {
	GitHubToken string
}{}

// Execute adds all child commands to the root command sets flags appropriately.
// This is called by main.main(). It only needs to happen once to the rootCmd.
func Execute() {
	if err := RootCmd.Execute(); err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
}

func init() {
	cobra.OnInitialize(initConfig)
}

// initConfig reads in config file and ENV variables if set.
func initConfig() {
	if os.Getenv("GITHUB_TOKEN") != "" {
		rootOpts.GitHubToken = os.Getenv("GITHUB_TOKEN")
	}
}


github/github.go
package github

import (
	"context"
	"net/http"
	"time"

	"github.com/google/go-github/v33/github"
	"golang.org/x/oauth2"
)

const (
	// default: 30, max: 100
	// https://developer.github.com/v3/#pagination
	perPage = 100
)

type Release struct {
	ArtifactURLs []string
	Author       string
	Body         string
	Commit       string
	CreatedAt    time.Time
	Name         string
	PublishedAt  time.Time
	URL          string
}

type Tag struct {
	Name    string
	Release *Release
}

type RepositoriesServiceInterface interface {
	GetReleaseByTag(ctx context.Context, owner, repo, tag string) (*github.RepositoryRelease, *github.Response, error)
	GetCommit(ctx context.Context, owner, repo, sha string) (*github.RepositoryCommit, *github.Response, error)
	ListReleases(ctx context.Context, owner, repo string, opt *github.ListOptions) ([]*github.RepositoryRelease, *github.Response, error)
	ListTags(ctx context.Context, owner string, repo string, opt *github.ListOptions) ([]*github.RepositoryTag, *github.Response, error)
}

type ClientInterface interface {
	DescribeRelease(ctx context.Context, owner, repo, tag string) (*Tag, error)
	ListTagsAndReleases(ctx context.Context, owner, repo string) ([]*Tag, error)
}

// Client represents a wrapper of GitHub API client
type Client struct {
	repositories RepositoriesServiceInterface
}

// NewClient creates new Client object
func NewClient(accessToken string) *Client {
	var hc *http.Client

	if accessToken == "" {
		hc = nil
	} else {
		ts := oauth2.StaticTokenSource(&oauth2.Token{
			AccessToken: accessToken,
		})
		hc = oauth2.NewClient(oauth2.NoContext, ts)
	}

	return &Client{
		repositories: github.NewClient(hc).Repositories,
	}
}

// DescribeRelease returns detail of the given release
func (c *Client) DescribeRelease(ctx context.Context, owner, repo, tag string) (*Tag, error) {
	release, err := c.getRelease(ctx, owner, repo, tag)
	if err != nil {
		return nil, err
	}

	commit, err := c.getTagCommit(ctx, owner, repo, tag)
	if err != nil {
		return nil, err
	}

	var body, name string

	if release.Body == nil {
		body = ""
	} else {
		body = *release.Body
	}

	if release.Name == nil {
		name = ""
	} else {
		name = *release.Name
	}

	artifactURLs := []string{}

	for _, asset := range release.Assets {
		artifactURLs = append(artifactURLs, *asset.BrowserDownloadURL)
	}

	createdAt := *release.CreatedAt
	publishedAt := *release.PublishedAt

	return &Tag{
		Name: *release.TagName,
		Release: &Release{
			ArtifactURLs: artifactURLs,
			Author:       *release.Author.Login,
			Body:         body,
			Commit:       *commit.SHA,
			CreatedAt: time.Date(
				createdAt.Year(),
				createdAt.Month(),
				createdAt.Day(),
				createdAt.Hour(),
				createdAt.Minute(),
				createdAt.Second(),
				createdAt.Nanosecond(),
				createdAt.Location(),
			),
			Name: name,
			PublishedAt: time.Date(
				publishedAt.Year(),
				publishedAt.Month(),
				publishedAt.Day(),
				publishedAt.Hour(),
				publishedAt.Minute(),
				publishedAt.Second(),
				publishedAt.Nanosecond(),
				publishedAt.Location(),
			),
			URL: *release.HTMLURL,
		},
	}, nil
}

func (c *Client) getRelease(ctx context.Context, owner, repo, tag string) (*github.RepositoryRelease, error) {
	release, _, err := c.repositories.GetReleaseByTag(ctx, owner, repo, tag)
	if err != nil {
		return nil, err
	}

	return release, nil
}

func (c *Client) getTagCommit(ctx context.Context, owner, repo, tag string) (*github.RepositoryCommit, error) {
	commit, _, err := c.repositories.GetCommit(ctx, owner, repo, tag)
	if err != nil {
		return nil, err
	}

	return commit, nil
}

// ListTagsAndReleases retrieves all tags and releases of the given repository
func (c *Client) ListTagsAndReleases(ctx context.Context, owner, repo string) ([]*Tag, error) {
	tags, err := c.listTags(ctx, owner, repo)
	if err != nil {
		return []*Tag{}, err
	}

	releases, err := c.listReleases(ctx, owner, repo)
	if err != nil {
		return []*Tag{}, err
	}

	releasesMap := map[string]*github.RepositoryRelease{}

	for _, release := range releases {
		releasesMap[*release.TagName] = release
	}

	ts := []*Tag{}

	for _, t := range tags {
		var tag *Tag

		if r, ok := releasesMap[*t.Name]; ok {
			var name string

			if r.Name == nil {
				name = ""
			} else {
				name = *r.Name
			}

			createdAt := *r.CreatedAt

			tag = &Tag{
				Name: *t.Name,
				Release: &Release{
					Name: name,
					CreatedAt: time.Date(
						createdAt.Year(),
						createdAt.Month(),
						createdAt.Day(),
						createdAt.Hour(),
						createdAt.Minute(),
						createdAt.Second(),
						createdAt.Nanosecond(),
						createdAt.Location(),
					),
				},
			}
		} else {
			tag = &Tag{
				Name: *t.Name,
			}
		}

		ts = append(ts, tag)
	}

	return ts, nil
}

// ListReleases lists all releases of the given repository
func (c *Client) listReleases(ctx context.Context, owner, repo string) ([]*github.RepositoryRelease, error) {
	allReleases := []*github.RepositoryRelease{}

	listOpts := &github.ListOptions{
		PerPage: perPage,
	}

	for {
		releases, resp, err := c.repositories.ListReleases(ctx, owner, repo, listOpts)
		if err != nil {
			return []*github.RepositoryRelease{}, err
		}

		allReleases = append(allReleases, releases...)

		if resp.NextPage == 0 {
			break
		}

		listOpts.Page = resp.NextPage
	}

	return allReleases, nil
}

// ListTags lists all tags of the given repository
func (c *Client) listTags(ctx context.Context, owner, repo string) ([]*github.RepositoryTag, error) {
	allTags := []*github.RepositoryTag{}

	listOpts := &github.ListOptions{
		PerPage: perPage,
	}

	for {
		tags, resp, err := c.repositories.ListTags(ctx, owner, repo, listOpts)
		if err != nil {
			return []*github.RepositoryTag{}, err
		}

		allTags = append(allTags, tags...)

		if resp.NextPage == 0 {
			break
		}

		listOpts.Page = resp.NextPage
	}

	return allTags, nil
}


cmd/list.go
package cmd

import (
	"context"
	"fmt"
	"io"
	"os"
	"strings"
	"text/tabwriter"
	"time"

	"github.com/dtan4/ghrls/github"
	"github.com/spf13/cobra"
)

// listCmd represents the list command
var listCmd = &cobra.Command{
	Use:   "list REPOSITORY",
	Short: "List releases",
	Long: `List releases

Example:

$ ghrls list kubernetes/kubernetes | head
TAG               TYPE           CREATEDAT                        NAME
v1.6.0-alpha.0    TAG
v1.5.3-beta.0     TAG
v1.5.2            TAG+RELEASE    2017-01-12 13:51:15 +0900 JST    v1.5.2
v1.5.2-beta.0     TAG
v1.5.1            TAG+RELEASE    2016-12-14 09:50:36 +0900 JST    v1.5.1
v1.5.1-beta.0     TAG
v1.5.0            TAG+RELEASE    2016-12-13 08:29:43 +0900 JST    v1.5.0
v1.5.0-beta.3     TAG+RELEASE    2016-12-09 06:52:35 +0900 JST    v1.5.0-beta.3
v1.5.0-beta.2     TAG+RELEASE    2016-11-25 07:29:04 +0900 JST    v1.5.0-beta.2
`,
	RunE: func(cmd *cobra.Command, args []string) error {
		timezone := time.Local
		client := github.NewClient(rootOpts.GitHubToken)

		return RunList(os.Stderr, os.Stdout, args, client, timezone)
	},
}

var (
	headers = []string{
		"TAG",
		"TYPE",
		"CREATEDAT",
		"NAME",
	}
)

func RunList(stdout, stderr io.Writer, args []string, client github.ClientInterface, timezone *time.Location) error {
	if len(args) != 1 {
		return fmt.Errorf("Please specify repository <user/name>.")
	}

	ss := strings.Split(args[0], "/")
	if len(ss) != 2 {
		return fmt.Errorf("Invalid repository name: %s", args[0])
	}
	owner, repo := ss[0], ss[1]

	ctx := context.Background()

	tags, err := client.ListTagsAndReleases(ctx, owner, repo)
	if err != nil {
		if strings.Contains(err.Error(), "404 Not Found") {
			return fmt.Errorf("%s/%s: not found", owner, repo)
		}
		return err
	}

	w := tabwriter.NewWriter(stdout, 0, 0, 4, ' ', 0)
	fmt.Fprintln(w, strings.Join(headers, "\t"))

	for _, tag := range tags {
		ss := []string{}

		if tag.Release != nil {
			ss = append(ss, tag.Name, "TAG+RELEASE", tag.Release.CreatedAt.In(timezone).String(), tag.Release.Name)
		} else {
			ss = append(ss, tag.Name, "TAG", "", "")
		}

		fmt.Fprintln(w, strings.Join(ss, "\t"))
	}

	w.Flush()

	return nil
}

func init() {
	RootCmd.AddCommand(listCmd)
}
